# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [3]:
import pandas as pd

# incorporar data frame 1
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

insurance_df = pd.read_csv(url)
insurance_df

# netejant data frame 1
def clean_column_names(df):
    df = df.rename(columns=lambda x: x.strip().lower().replace(' ', '_'))
    df = df.rename(columns={'st': 'state'})
    df = df.rename(columns={'customer': 'customer_id'})
    df = df.rename(columns={'education': 'education_level'})
    df = df.rename(columns={'monthly_premium_auto': 'monthly_auto_premium'})
    df = df.rename(columns={'number_of_open_complaints': 'open_complaints_count'})
    return df

# comprobant que estigui neta
insurance_df = clean_column_names(insurance_df)
insurance_df

# eliminant totes les files que son Nan values de la primera data frame
insurance_df = insurance_df.drop(insurance_df.tail(2937).index)
insurance_df

# incortorant data frame 2
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv"

insurance_df_2 = pd.read_csv(url)
insurance_df_2.head()

# cleaning data frame 2
def clean_column_names(df):
    df = df.rename(columns=lambda x: x.strip().lower().replace(' ', '_'))
    df = df.rename(columns={'st': 'state'})
    df = df.rename(columns={'customer': 'customer_id'})
    df = df.rename(columns={'education': 'education_level'})
    df = df.rename(columns={'monthly_premium_auto': 'monthly_auto_premium'})
    df = df.rename(columns={'number_of_open_complaints': 'open_complaints_count'})
    return df

# comprobant que estigui neta
insurance_df_2 = clean_column_names(insurance_df_2)
insurance_df_2

# incorporant 3
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv"

insurance_df_3 = pd.read_csv(url)
insurance_df_3

# netajant data frame 3
def clean_column_names(df):
    df = df.rename(columns=lambda x: x.strip().lower().replace(' ', '_'))
    df = df.rename(columns={'st': 'state'})
    df = df.rename(columns={'customer': 'customer_id'})
    df = df.rename(columns={'education': 'education_level'})
    df = df.rename(columns={'monthly_premium_auto': 'monthly_auto_premium'})
    df = df.rename(columns={'number_of_open_complaints': 'open_complaints_count'})
    return df

insurance_df_3 = clean_column_names(insurance_df_3)
insurance_df_3

# unint les tres taules
df_total = pd.concat([insurance_df, insurance_df_2, insurance_df_3], ignore_index=True)
df_total


# Cleaning invalid values
# Gender
df_total['gender'].unique()

df_total['gender'] = df_total['gender'].str.strip().str.upper()  # Remove spaces & unify case
df_total['gender'] = df_total['gender'].replace({
    'FEMALE': 'F',
    'FEMAL': 'F',
    'MALE': 'M'
})

# State
df_total['state'].unique()

state_map = {
    'AZ': 'Arizona',
    'Cali': 'California',
    'CALI': 'California',
    'WA': 'Washington',
    'NV':'Nevada',
    'OR':'Oregon'
}

df_total['state'] = df_total['state'].replace(state_map)
df_total['state'] = df_total['state'].str.strip()
df_total['state'].unique()

# Education
df_total['education_level'].unique()

education_map = {
    'Bachelors': 'Bachelor',
}

df_total['education_level'] = df_total['education_level'].replace(education_map)
df_total['education_level'].unique()

# Customer Lifetime value - deliting character %
df_total['customer_lifetime_value'] = (df_total['customer_lifetime_value'].astype(str).str.replace('%', '', regex=False))
df_total['customer_lifetime_value'] = pd.to_numeric(df_total['customer_lifetime_value'], errors='coerce')

# Vehicle Class
df_total['vehicle_class'] = df_total['vehicle_class'].replace({
    'Sports Car': 'Luxury',
    'Luxury SUV': 'Luxury',
    'Luxury Car': 'Luxury'
})
df_total['vehicle_class'].unique()

# 3. Formatting data types
# Open complaints counts
#Clearing data
df_total['open_complaints_count'] = df_total['open_complaints_count'].apply(lambda x: x.split('/')[1] if isinstance(x, str) and '/' in x else x)
df_total['open_complaints_count'].unique()

# Chaning to intiger
df_total['open_complaints_count'] = pd.to_numeric(df_total['open_complaints_count'], errors='coerce').astype('Int64')
df_total['open_complaints_count'].dtype

# comprovant com queda la taula
df_total

customer_id       state gender       education_level  \
0        RB50392  Washington    NaN                Master   
1        QZ44356     Arizona      F              Bachelor   
2        AI49188      Nevada      F              Bachelor   
3        WW63253  California      M              Bachelor   
4        GA49547  Washington      M  High School or Below   
...          ...         ...    ...                   ...   
9132     LA72316  California      M              Bachelor   
9133     PK87824  California      F               College   
9134     TD14365  California      M              Bachelor   
9135     UP19263  California      M               College   
9136     Y167826  California      M               College   

      customer_lifetime_value   income  monthly_auto_premium  \
0                         NaN      0.0                1000.0   
1                6.979536e+05      0.0                  94.0   
2                1.288743e+06  48767.0                 108.0   
3                7.645862e+05      0.0                 106.0   
4                5.363077e+05  36357.0                  68.0   
...                       ...      ...                   ...   
9132             2.340599e+04  71941.0                  73.0   
9133             3.096511e+03  21604.0                  79.0   
9134             8.163890e+03      0.0                  85.0   
9135             7.524442e+03  21941.0                  96.0   
9136             2.611837e+03      0.0                  77.0   

      open_complaints_count     policy_type  vehicle_class  total_claim_amount  
0                         0   Personal Auto  Four-Door Car            2.704934  
1                         0   Personal Auto  Four-Door Car         1131.464935  
2                         0   Personal Auto   Two-Door Car          566.472247  
3                         0  Corporate Auto            SUV          529.881344  
4                         0   Personal Auto  Four-Door Car           17.269323  
...                     ...             ...            ...                 ...  
9132                      0   Personal Auto  Four-Door Car          198.234764  
9133                      0  Corporate Auto  Four-Door Car          379.200000  
9134                      3  Corporate Auto  Four-Door Car          790.784983  
9135                      0   Personal Auto  Four-Door Car          691.200000  
9136                      0  Corporate Auto   Two-Door Car          369.600000  

[9137 rows x 11 columns]

In [4]:
df_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9137 entries, 0 to 9136
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              9137 non-null   object 
 1   state                    9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education_level          9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   float64
 5   income                   9137 non-null   float64
 6   monthly_auto_premium     9137 non-null   float64
 7   open_complaints_count    9137 non-null   Int64  
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: Int64(1), float64(4), object(6)
memory usage: 794.3+ KB


In [5]:
df_total.isna().sum()

customer_id                  0
state                        0
gender                     122
education_level              0
customer_lifetime_value      7
income                       0
monthly_auto_premium         0
open_complaints_count        0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [9]:
df_total['gender'].mode()

0    F
Name: gender, dtype: object

In [14]:
mode_val_gender = df_total["gender"].mode().iloc[0]
df_total["gender"] = df_total["gender"].fillna(mode_val_gender)

In [13]:
median_val_clv = df_total["customer_lifetime_value"].median()
df_total["customer_lifetime_value"] = df_total["customer_lifetime_value"].fillna(median_val_clv)

In [15]:
df_total.isna().sum()

customer_id                0
state                      0
gender                     0
education_level            0
customer_lifetime_value    0
income                     0
monthly_auto_premium       0
open_complaints_count      0
policy_type                0
vehicle_class              0
total_claim_amount         0
dtype: int64

In [16]:
insurance_df.duplicated().sum()

np.int64(0)

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [10]:
# incorporar la nova data frame
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv"

marketing_customer = pd.read_csv(url)
marketing_customer

unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877  California              6857.519928       No   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College        2011-02-18         Employed      M  ...   
1         Basic   College        2011-01-18       Unemployed      F  ...   
2         Basic  Bachelor        2011-02-10         Employed      M  ...   
3      Extended   College        2011-01-11         Employed      M  ...   
4       Premium  Bachelor        2011-01-17    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor        2011-01-19       Unemployed      F  ...   
10906     Basic   College        2011-01-06         Employed      F  ...   
10907  Extended  Bachelor        2011-02-06         Employed      F  ...   
10908   Premium   College        2011-02-13         Employed      F  ...   
10909     Basic  Bachelor        2011-01-08       Unemployed      M  ...   

       number_of_policies     policy_type        policy  renew_offer_type  \
0                       9  Corporate Auto  Corporate L3            Offer3   
1                       1   Personal Auto   Personal L3            Offer4   
2                       2   Personal Auto   Personal L3            Offer3   
3                       2  Corporate Auto  Corporate L3            Offer2   
4                       7   Personal Auto   Personal L2            Offer1   
...                   ...             ...           ...               ...   
10905                   7   Personal Auto   Personal L1            Offer3   
10906                   6   Personal Auto   Personal L3            Offer2   
10907                   2  Corporate Auto  Corporate L3            Offer1   
10908                   6   Personal Auto   Personal L1            Offer1   
10909                   3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0              Agent          292.800000  Four-Door Car       Medsize   
1        Call Center          744.924331  Four-Door Car       Medsize   
2        Call Center          480.000000            SUV       Medsize   
3             Branch          484.013411  Four-Door Car       Medsize   
4             Branch          707.925645  Four-Door Car       Medsize   
...              ...                 ...            ...           ...   
10905            Web         1214.400000     Luxury Car       Medsize   
10906         Branch          273.018929  Four-Door Car       Medsize   
10907            Web          381.306996     Luxury SUV       Medsize   
10908         Branch          618.288849            SUV       Medsize   
10909            Web         1021.719397            SUV       Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 27 columns]

In [11]:
# eliminem columnes que no ens son utils

marketing_customer = marketing_customer.drop(columns=['unnamed:_0'])

In [12]:
# comprobem els tipos de dates que tenim 

marketing_customer.dtypes

customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
vehicle_type    

In [13]:
marketing_customer['sales_channel'].unique()

array(['Agent', 'Call Center', 'Branch', 'Web'], dtype=object)

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [21]:
# fem pivot amb sales i revenue

pivot_revenue = pd.pivot_table(marketing_customer,
    values = 'total_claim_amount',
    index = 'sales_channel',
    aggfunc = 'sum'
).round(2).sort_values(by='total_claim_amount', ascending=False)

pivot_revenue

total_claim_amount
sales_channel                    
Agent                  1810226.82
Branch                 1301204.00
Call Center             926600.82
Web                     706600.04

# conclusions:
- The Agents channel records the highest total claim amount (≈ 1.81M), positioning it as the top-performing sales channel in terms of monetary value.
- Branches show solid performance with ≈ 1.30M.
- The Call Center (≈ 0.93M), indicating potential for improvement or targeted optimization strategies.
- The Web channel has the lowest total claim amount (≈ 0.71M), which may reflect lower engagement.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [26]:
pivot_table2 = pd.pivot_table(marketing_customer,
    index = ['gender', 'education'],
    values ='customer_lifetime_value', # columna per analitzar
    aggfunc = 'mean'                # funció agregada: avg             
).round(2)                           # arodonim a 2 decimals

pivot_table2

customer_lifetime_value
gender education                                    
F      Bachelor                              7874.27
       College                               7748.82
       Doctor                                7328.51
       High School or Below                  8675.22
       Master                                8157.05
M      Bachelor                              7703.60
       College                               8052.46
       Doctor                                7415.33
       High School or Below                  8149.69
       Master                                8168.83

# conclusions: 
- Highest CLV segments: Women with High School or Below (8,675.22) and Master (8,157.05) or Men with Master (8168.83) or College degrees (8052.46), indicating these groups generate the most value.
- Gender trends: Men show consistently high CLV in Master and College categories, while women exhibit an unexpected peak in High School or Below.
- Insights: Target retention and upselling efforts to Master, College, and high-CLV female segments. Investigate the drivers behind the high CLV in women with lower education for potential replication across other segments.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [31]:
pivot_table3 = pd.pivot_table(marketing_customer,
    index = ['policy_type', 'month'],
    values = 'number_of_open_complaints',
    aggfunc = 'sum'
).reset_index().sort_values(by='number_of_open_complaints', ascending=False)
pivot_table3

policy_type  month  number_of_open_complaints
2   Personal Auto      1                1727.605722
3   Personal Auto      2                1453.684441
0  Corporate Auto      1                 443.434952
1  Corporate Auto      2                 385.208135
5    Special Auto      2                  95.226817
4    Special Auto      1                  87.074049